kimchi/src/circuits/expr.rs

```diff
@@ -540,8 +551,148 @@ pub enum Expr<C> {
     UnnormalizedLagrangeBasis(i32),
     Pow(Box<Expr<C>>, u64),
     Cache(CacheId, Box<Expr<C>>),
-    /// Expression is conditional on the given feature flag, returns 0 if disabled.
-    EnabledIf(FeatureFlag, Box<Expr<C>>),
+    /// If the feature flag is enabled, return the first expression; otherwise, return the second.
+    IfFeature(FeatureFlag, Box<Expr<C>>, Box<Expr<C>>),
+}
+
+impl<C: Zero + One + Neg<Output = C> + PartialEq + Clone> Expr<C> {
+    fn apply_feature_flags_inner(&self, features: &FeatureFlags) -> (Expr<C>, bool) {
+        use Expr::*;
+        match self {
+            Constant(_) | Cell(_) | VanishesOnLast4Rows | UnnormalizedLagrangeBasis(_) => {
+                (self.clone(), false)
+            }
+            Double(c) => {
+                let (c_reduced, reduce_further) = c.apply_feature_flags_inner(features);
+                if reduce_further && c_reduced.is_zero() {
+                    (Expr::zero(), true)
+                } else {
+                    (Double(Box::new(c_reduced)), false)
+                }
+            }
+            Square(c) => {
+                let (c_reduced, reduce_further) = c.apply_feature_flags_inner(features);
+                if reduce_further && (c_reduced.is_zero() || c_reduced.is_one()) {
+                    (c_reduced, true)
+                } else {
+                    (Square(Box::new(c_reduced)), false)
+                }
+            }
+            BinOp(op, c1, c2) => {
+                let (c1_reduced, reduce_further1) = c1.apply_feature_flags_inner(features);
+                let (c2_reduced, reduce_further2) = c2.apply_feature_flags_inner(features);
+                match op {
+                    Op2::Add => {
+                        if reduce_further1 && c1_reduced.is_zero() {
+                            if reduce_further2 && c2_reduced.is_zero() {
```

Comment:
Why can the zero expression be reduced further? Meaning, why is this a true? I guess I am not seeing what the bool here exactly means